ejb3/src/main/java/org/jboss/as/ejb3/subsystem/TransactionNamespaceAccessControlService.java

```diff
@@ -0,0 +1,126 @@
+ * by the @authors tag. See the copyright.txt in the distribution for a
+ * full listing of individual contributors.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.jboss.as.ejb3.subsystem;
+
+import org.jboss.as.ejb3.component.allowedmethods.AllowedMethodsInformation;
+import org.jboss.as.ejb3.component.allowedmethods.MethodType;
+import org.jboss.as.naming.InServerInitialContextFactoryService;
+import org.jboss.as.naming.LookupInterceptor;
+import org.jboss.msc.service.Service;
+import org.jboss.msc.service.ServiceName;
+import org.jboss.msc.service.StartContext;
+import org.jboss.msc.service.StartException;
+import org.jboss.msc.service.StopContext;
+import org.jboss.msc.value.InjectedValue;
+
+import javax.naming.InvalidNameException;
+import javax.naming.Name;
+import javax.naming.NamingException;
+
+/**
+ * @author <a href="mailto:tadamski@redhat.com">Tomasz Adamski</a>
+ */
+class TransactionNamespaceAccessControlService implements Service<Void> {
+
+    private static final String TRANSACTION_NAMESPACE = "txn";
+    private static final String REMOTE_USER_TRANSACTION = "RemoteUserTransaction";
+    private static final String LOCAL_USER_TRANSACTION = "LocalUserTransaction";
+
+    public static final ServiceName SERVICE_NAME = ServiceName.JBOSS.append("ejb3", "TransactionNamespaceAccessControlService");
+
+    private final InjectedValue<InServerInitialContextFactoryService> initialContextFactory = new InjectedValue<>();
+
+    private LookupInterceptor userTransactionLookupInterceptor = new LookupInterceptor() {
+
+        @Override
+        public void aroundLookup(Name name) throws NamingException {
+            ReparsedName reparsedName = reparse(name);
+            if (reparsedName.getUrlScheme().equals(TRANSACTION_NAMESPACE) &&
+                    (reparsedName.getName().toString().equals(REMOTE_USER_TRANSACTION) || reparsedName.getName().toString().equals(LOCAL_USER_TRANSACTION))) {
+                try {
+                    AllowedMethodsInformation.checkAllowed(MethodType.GET_USER_TRANSACTION);
+                } catch (IllegalStateException e) {
+                    throw new NamingException(e.getMessage());
+                }
+            }
+        }
```

Comment:
This is executed on the request path so the algorithm needs to be highly optimized with minimal object allocations. ReparsedName exists to pass a bit of state from a private method so the calling method can use that state in if checks. Instead this could be 'if (requiresCheckAllowed(name)) {' with requiresCheckAllowed(Name name) doing an optimized check.

Reply:
fixed